### ISSUE TITLE
Rope renames variable incorrectly

### ISSUE THREAD

**Issue by hbusul** [python-rope/rope#288]
I've come across with this issue while using Rope as a library.

Python version: 3.7.6
OS: Ubuntu 18.04
Rope Version: 0.16.0

I also noticed this bug does not occur on Rope version 0.14.0

## How to reproduce the error
1. Create the following file structure
```
app
├── run.py
└── my_app
|   ├── __init__.py
|   └── apis
|      └── __init__.py
|      └── my_namespace.py
```

```
#app/run.py
from my_app import app

if __name__ == '__main__':
    app.run(host='0.0.0.0')

#app/my_app/__init__.py
from flask import Flask
from flask_httpauth import HTTPBasicAuth, HTTPTokenAuth, MultiAuth

app = Flask(__name__)
auth = MultiAuth(HTTPBasicAuth(), HTTPTokenAuth('Bearer'))

from my_app.apis import api  # noqa
api.init_app(app)

#app/my_app/apis/__init__.py
from flask_restplus import Api
from apis.my_namespace import api as my_api


api = Api()
api.add_namespace(my_api)


#app/my_app/apis/my_namespace.py
from flask_restplus import Namespace, Resource
from my_app import auth

api = Namespace('somenamespace')


class MyClassA(Resource):
    @auth.login_required
    def delete(self):
        my_var = f"some value"


class MyClassB(Resource):
    @auth.login_required
    def get(self):
        pass


api.add_resource(MyClassA, '/class-a')
api.add_resource(MyClassB, "/class-b")
```

2. Create the script for renaming `auth` variable in `app/my_app/__init__.py`.
This script should be at the same level with app folder. Name it `renamer.py`.
```
from rope.base import libutils, project
from rope.refactor.rename import Rename


def main():
    my_project = project.Project("app")
    file = "app/my_app/__init__.py"
    file_as_res = libutils.path_to_resource(my_project, file)
    declaration = 115
    change_set = Rename(my_project, file_as_res,
                        declaration).get_changes("f162b77e780b")
    my_project.do(change_set)
    my_project.close()


if __name__ == "__main__":
    main()
```

3. Call that script `python renamer.py`

## Expected Result
In the `app/my_app/apis/my_namespace.py` it should look like:
```
from flask_restplus import Namespace, Resource
from my_app import f162b77e780b

api = Namespace('somenamespace')


class MyClassA(Resource):
    @f162b77e780b.login_required
    def delete(self):
        my_var = f"some value"


class MyClassB(Resource):
    @f162b77e780b.login_required
    def get(self):
        pass


api.add_resource(MyClassA, '/class-a')
api.add_resource(MyClassB, "/class-b")
```
## Observed Result
```
from flask_restplus import Namespace, Resource
from my_app import f162b77e780b

api = Namespace('somenamespace')


class MyClassA(Resource):
    @f162b77e780b.login_required
    def delete(self):
        my_var = f"some value"


class MyClassB(Resource):
    @af162b77e780buth.login_required # <<<<---- notice here 
    def get(self):
        pass


api.add_resource(MyClassA, '/class-a')
api.add_resource(MyClassB, "/class-b")

```

**Comment by pappasam** [comment]
Piggy backing on this issue, renaming doesn't work at all for me on the latest version of rope with `coc-python` on Neovim. Version 0.14.0 seems to actually perform a rename, but the latest version apparently does nothing. Did the public API change dramatically between 0.14 and 0.16?

**Comment by mcepl** [comment]
I don't think there should be any change in the public API. Looking at https://github.com/python-rope/rope/compare/0.14.0...0.16.0 it looks really innocent. Would you be able to bisect the issue to some particular commit, please?

**Comment by hbusul** [comment]
It works with commit [`975a4800f4ebab07fc779fe85700db3466a4168b`](https://github.com/python-rope/rope/commit/975a4800f4ebab07fc779fe85700db3466a4168b) `Fix return value of __enter__ for builtin File` however it seems like [`5c3da463940b44c8796f2a9edccecd4bd4e2ecc7`](https://github.com/python-rope/rope/commit/5c3da463940b44c8796f2a9edccecd4bd4e2ecc7) `Rename variables in f-strings` seems to break it. Given that in the case there is `my_var = f"some value"`, I guess that's the reason. Normally I was using f string with parameters but I got rid of them to make case smaller.

**Comment by pappasam** [comment]
tl;dr my hunch: something is wrong with `base.codeanalyze.get_string_pattern`.

I think most problems are with 5c3da463 and the changes to `base.codeanalyze.py`.

`base.codeanalyze.get_string_pattern` is called by `refactor.occurences._TextualFinder`,  which is called by `refactor.occurences.create_finder` which calls `refactor.occurences.Finder` which is originally called by `refactor.rename.rename_in_module` which is invoked by both:

* `refactor.rename.ChangeOccurences.get_changes` AND
* `refactor.rename.Rename.get_changes`

**Comment by hbusul** [comment]
I minified my test code to
```
auth = 8
my_var = f"some value {auth}"
print(auth)
other_val = "some other"
```
I realized the pattern finds offsets correctly, however `_RealFinder` was __unable__ to find word start(`_find_word_start`) and word end(`_find_word_end`)  correctly. When I investigated further I realized it was due to wrong simplification of the code.
This code
```
auth = 8
my_var = f"some value {auth}"
print(auth)
other_val = "some other"
```
is being simplified to
```
auth = 8
my_var = f"some value {auth}"                         "some other"
```
It's my first time studying the source code but if I got it right, simplification is applied to make process easier and there is a function called `ignored_regions` and it `Return ignored regions like strings and comments in source ` . Regular expression is incorrect. As @pappasam pointed out, `base.codeanalyze.get_string_pattern` is also used in `ignored_regions`. I got the relevant part of the regular expression that is 

`(?<![fF])"(\.|\n|[^"\\])*"`

However it is matching(also you can try it [here](https://regexr.com/4ttbv)),

```
"
print(auth)
other_val = "
```

and due to that simplification is incorrect. I don't know if strings should be ignored after all but hope this helps!

**Comment by mcepl** [comment]
I am sorry, I am swamped by my real work at the moment, but this is truly a great help! @soupytwist any hope for helping here?

**Comment by mcepl** [comment]
@hbusul What’s the status of this ticket after #289 has been merged? Is this fixed?

**Comment by hbusul** [comment]
Yes, it solved the issue. Sorry for forgetting to close the issue. When will this fix will be available in Python Package Index?

**Comment by mcepl** [comment]
> Yes, it solved the issue. Sorry for forgetting to close the issue. When will this fix will be available in Python Package Index?

I really need to get those Python 3.8-related bugs fixed. So, I guess, this milestone https://github.com/python-rope/rope/milestone/3 .
